keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+    * logr implementation to be used by new logs (`json` loggers for `json` log format and klogr logger for legacy format)
+* For each package in Kubernetes source code define a logging module to which it will belong. Each logging module can span across multiple packages.
+* For each logging module define two variables with logr implementations to be used for existing and new logs initialized with the name of the module. Those should only be used until stage 2 of the migration is fully completed.
+* Implement a package implementing klog API which would check a global variable and either passthrough call to klog (with a proper deep value handing) or do nothing. This global variable should be initialized based on `--logging-format` value.
+* Kubernetes and all of its dependencies should be upgraded to klog v2.
+
+#### Automatic migration
+
+1. If `json` log format is selected klog v2 library should be initialized with json logr implementation so that all logs produced by Kubernetes or its dependencies using klog API would be redirected to logr.
+1. For all golang packages in Kubernetes following changes should be performed:
+    1. All imports and usages of klog package should be replaced with newly created klog wrapper.
+    1. For all calls to klog.V().Info*(), klog.Info*(), klog.Warning*(), klog.Error*(), klog.Fatal*() or klog.Exit*() corresponding logr invocation should be added just after those.
+
+This should be done automatically for the whole source code of the Kubernetes by a tool which would analyze the source code and edit it accordingly. The result of the automated migration should be verified and where necessary cleaned up manually.
+Generated PRs should be divided by directories and send for review.
+
+In case where constant string is passed as a message argument to logr.Info() or logr.Error() it should be replaced with a normalized constant string produced by removing all formatting placeholders like %s, %q and %v and removing any trailing non alphanumeric characters from its end:
+
+"Update ready status of pods on node [%v]"	-> "Update ready status of pods on node"
+"Failed to update status for pod %q: %v"	-> "Failed to update status for pod"
+
+#### Stage 2 / Beta
+
+Logging context propagation
+
+1. New `k8s.io/component-base/kontext` package should be implemented which would provide functionalities related to Context handling in Kubernetes:
+    * Attach given logr.Logger instance to existing Context instance.
+    * Get logr.Logger attached to given Context.
+    * Attach new Kubernetes object to Context and attached logr.Logger.
+    * Get Kubernetes object attached to given Context.
+1. All entry points to the application like main() function and HTTP request handlers should be identified. In each of those entry points a new Context should be created if it does not exist already and a module logger should be associated with this context.
```

Comment:
this is helpful, for things like controller manager or kubelet, is there guidance on when objects would be added to and/or removed from context?

Reply:
We planned context object to be immutable and therefore we have not planned to add any option to remove an object from a context. Instead of removing an object from a context context object itself should be simply discarded.
When it comes to adding objects to context in my mind this should basically follow a scoping of variables containing those objects in the source code. So as soon as we have some Kubernetes object assigned to some variable for the first time we should create a new context by calling `kontext.WithObject()` and this context should be scoped the same as this variable.

Reply:
if i am not mistaken, the context copies the objects that get added, not sure the cost of those copies depending on how often we add objects, but just recording so my future self remembers what i had looked up ;-)